### PR TITLE
chore(tests): Add `fork_test` helper; refactor API tests

### LINKED ISSUE #4925: Update problematic `api` tests to run in different processes
API tests are currently failing in #4886 due to shared global state. This isn't simple to isolate currently, due to the global nature of tracing and metrics.

A workaround could be using [rusty-fork](https://crates.io/crates/rusty-fork), but known disadvantages are:

- It doesn't currently [support async tests](https://github.com/AltSysrq/rusty-fork/issues/9).
- Wrapping tests in a `rusty_fork_test!` macro disables code analysis in IDEs, making it harder to write/maintain tests.

Need to investigate whether there's an option to combine with `#[tokio::test]` and/or create a thin wrapping function which executes the test in a separate runtime, so tests themselves can be written as regular functions.

### PR DESCRIPTION
**Pull request by leebenson**
Adds a `tests::support::fork_test()` helper, for spawning tests in an independent process. This is specifically helpful for API tests that observe and mutate globally instantiated metrics/tracing.

The `#[ignore]` attribute has been removed from subscription tests; this should unblock all tests for CI.

Closes #4925

(Currently a draft until I've confirmed CI is green.)

### CODE REVIEW

File: tests/api.rs
Comment (leebenson) [inline tests/api.rs:37]: Dropped the `sync::Once` wrapper, since tests that use it now run in isolated processes.

File: tests/api.rs
Comment (leebenson) [inline tests/api.rs:53]: A convenience for calling `fork_test`, with initialised metrics.

File: tests/api.rs
Comment (leebenson) [inline tests/api.rs:281]: This test already existed. I've just renamed it to fit the layout of the other tests.

File: tests/api.rs
Comment (leebenson) [inline tests/api.rs:451]: In order for tests to be identified correctly across processes, the canonical test name should be provided here in the format of `mod::function`.

I had originally tried helper macros such as [function_name!](https://crates.io/crates/function_name) to make this play better with refactoring, but it added unnecessary boilerplate for what is a fairly simple copy/paste job.

File: tests/support/mod.rs
Comment (leebenson) [inline tests/support/mod.rs:411]: rusty_fork provides a [`wait_with_output`](https://docs.rs/rusty-fork/0.3.0/rusty_fork/struct.ChildWrapper.html#method.wait_with_output) method, which provides a direct buffer to stdout via its `.stdout`, sidestepping the need to copy from the captured `File`.

However, child is shared as `&mut child` to this closure, and `wait_with_output` returns an owned value, so it wasn't clear how this could be used.

File: tests/api.rs
Comment (sghall) [reply tests/api.rs:373]: Yeah, think we can work with that.  Just have to be careful.  If you put `tests::some_random_name` in there it will happily just pass.

**Comment by sghall** [comment]
Tried this out locally and works as expected.  I think this is a reasonable way to get these tests going again.  As @jszwedko mentioned https://github.com/timberio/vector/issues/4925#issuecomment-724072768 I think it would be interesting to consider refactoring the global state pieces in the longer term.  That's a bigger project for sure.

**Review by sghall — APPROVED** [review]

**Comment by leebenson** [comment]
@sghall, I noticed the component link test can sometimes fail on master currently. Took 10-15 goes for it to show up, so seems like perhaps a subtle race condition. Haven't been able to replicate with this PR so far, so hopefully fixed 🤞

File: tests/api.rs
Comment (leebenson) [reply tests/api.rs:373]: Yeah, it's a shame there's no `nameof()` equivalent to keep these in sync. There are macros, but they wind up adding more LoC and still require developer buy-in, so I didn't feel they were adding much.

I'd have preferred an approach where `fork_test` knew the calling function and we could avoid this string name altogether, but I'm guessing any concept of 'caller' has been elided by the time this runs. I did poke around for a macro that might be able to fill that in based on usage, but turned up empty handed.

For now, I think this is fine, but it does rely on some manual checking.

**Comment by binarylogic** [comment]
Noting, I'd like to get @lukesteensen's input before merging this.

File: tests/api.rs
Comment (jszwedko) [reply tests/api.rs:373]: I wonder if we should structure `metrics_test` as macro? Like `#[metrics_test]`. I think that would resolve this issue as well.

**Review by jszwedko — COMMENTED** [review]
Assuming we want to go this fork route (I defer to @lukesteensen for that), I think this looks good. Nice work!

I am wondering if it wouldn't be cleaner to, instead of a `metrics_test` function, have a `#[metrics_test]` macro (similar to `tokio::test`) that would handle the setup.

File: tests/api.rs
Comment (leebenson) [reply tests/api.rs:373]: @jszwedko, love it! I think it'd need to be a proc-macro though, and be hosted in another crate. Open to moving forking to another lib for this purpose. Over to @lukesteensen for thoughts.

Comment (lukesteensen) [reply tests/api.rs:373]: It does seem like a proc macro with that helper crate could probably clean this up a bit, but honestly, I'm not sure it'd be worthwhile unless we plan to keep writing a lot more of these particular tests. The whole situation (global state, forking, etc) makes me a bit sad, so it'd be nice to limit the number of these kinds of tests we need to write and cover any additional ongoing testing needs with a type of test that's simpler and more direct.

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**
